File: app/src/main/java/com/viviose/musicscratchpad/Note.java
package com.viviose.musicscratchpad;

import android.util.DisplayMetrics;

import java.util.ArrayList;

/**
 * Created by Patrick on 2/3/2016.
 */
public class Note{
    public float x;
    public float y;
    float interval = DensityMetrics.spaceHeight;
    float subInt = DensityMetrics.spaceHeight / 2;
    float subSubInt = subInt / 2;
    NoteName name;
    int octave = Octave.octave;
    public enum NoteName{
        c,
        cs,
        d,
        ds,
        e,
        f,
        fs,
        g,
        gs,
        a,
        as,
        b
    }
    private NoteName[] altoStandardNotes =      {NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b};
    private NoteName[] trebleStandardNotes =    {NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a};
    private NoteName[] bassStandardNotes =      {NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c};

    public class NotePosn{
        NoteName noteName;
        float yLower;
        float yUpper;
        int octave;
        public NotePosn(NoteName n, float yL, float yU, int oct){
            noteName = n;
            yLower = yL;
            yUpper = yU;
            octave = oct;
        }
    }

    private ArrayList ALTO_POSN = new ArrayList();

    private ArrayList<NotePosn> getNotePosn(float y, Clef clef){
        NoteName[] iterArray = new NoteName[9];
        ArrayList<NotePosn> res = new ArrayList();
        switch (clef){
            case ALTO:
                iterArray = altoStandardNotes;
                res = ALTO_POSN;
                break;
            case TREBLE:
                iterArray = trebleStandardNotes;
                res = ALTO_POSN;
                Octave.octave = 5;
                break;
            case BASS:
                iterArray = bassStandardNotes;
                res = ALTO_POSN;
                Octave.octave = 4;
        }
        for (int i = 13; i > -1; i--){
            float yL = (interval * 8 - subInt) - subInt * i + DensityMetrics.getToolbarHeight();
            float yU = (interval * 8) - subInt * i + DensityMetrics.getToolbarHeight();
            int o = octave;
            if (clef == Clef.ALTO){
                if (i < 7){
                    o = octave - 1;
                }
            }else if (clef == Clef.TREBLE){
                if (i < 8){

                    o = octave - 1;
                }
            }else if (clef == Clef.BASS) {
                if (i < 6) {
                    o = octave - 2;
                }else if ( i < 13 && i >= 6){
                    o = octave - 1;
                }

            }

            res.add(new NotePosn(iterArray[i], yL, yU, o));
        }
        return res;
    }

    private NoteName getNoteFromPosn(float y, Clef clef){
        ArrayList<NotePosn> noteList = getNotePosn(y, clef);
        for (int i = 0; i < noteList.size(); i++) {
            if (y - DensityMetrics.getToolbarHeight() > noteList.get(i).yLower && y - DensityMetrics.getToolbarHeight() <= noteList.get(i).yUpper){
                octave = noteList.get(i).octave;
                return noteList.get(i).noteName;
            }
        }
        octave += 1;
        return NoteName.c;
    }

    private NoteName keyAccidental(NoteName note){
        if (Key.COUNT == 0){
            return note;
        }
        if (Key.SHARP){
            NoteName[][] sharps = { {NoteName.f, NoteName.fs},
                                    {NoteName.c, NoteName.cs},
                                    {NoteName.g, NoteName.gs},
                                    {NoteName.d, NoteName.ds},
                                    {NoteName.a, NoteName.as},
                                    {NoteName.e, NoteName.f},
                                    {NoteName.b, NoteName.c}};
            for (int i = 0; i < sharps.length; i++){
                if (note == sharps[i][0]){
                    return sharps[i][1];
                }
            }
        }
        return note;
    }




    public Note(float xC, float yC){
        x = xC;
        y = snapNoteY(yC);
        name = getNoteFromPosn(yC, ClefSetting.clef);

    }


//Change this to a rounding statement
    private float snapNoteY(float y) {

        float snapY = 0;
        for (int i = 1; i < 9; i++) {
            if (interval * i - subSubInt < y && interval * i + subSubInt >= y) {
                snapY = interval * i;
            }else if(interval * i + subInt - subSubInt < y && interval * i + subInt + subSubInt >= y){
                snapY = interval * i + subInt;
            }
        }
        return snapY;
    }


}

File: app/src/main/java/com/viviose/musicscratchpad/Note.kt
package com.viviose.musicscratchpad

import java.util.*

/**
 * Created by Patrick on 2/3/2016.
 */
class Note(var x: Float, yC: Float, acc: Int) {
    var y: Float = 0.toFloat()
    internal var interval = DensityMetrics.spaceHeight
    internal var subInt = interval / 2
    internal var subSubInt = subInt / 2
    var name: NoteName?
    var rhythm: Double = 0.toDouble()
    var octave = Octave.octave
    internal var accidental: Int = acc
    internal var SHARPS = arrayOf(arrayOf(NoteName.f, NoteName.fs), arrayOf(NoteName.c, NoteName.cs), arrayOf(NoteName.g, NoteName.gs), arrayOf(NoteName.d, NoteName.ds), arrayOf(NoteName.a, NoteName.`as`), arrayOf(NoteName.e, NoteName.f), arrayOf(NoteName.b, NoteName.c))
    internal var FLATS = arrayOf(arrayOf(NoteName.b, NoteName.`as`), arrayOf(NoteName.e, NoteName.ds), arrayOf(NoteName.a, NoteName.gs), arrayOf(NoteName.d, NoteName.cs), arrayOf(NoteName.g, NoteName.fs), arrayOf(NoteName.c, NoteName.b), arrayOf(NoteName.f, NoteName.e))

    enum class NoteName {
        c,
        cs,
        d,
        ds,
        e,
        f,
        fs,
        g,
        gs,
        a,
        `as`,
        b
    }

    private val altoStandardNotes: Array<NoteName?> = arrayOf(NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c)
    private val trebleStandardNotes: Array<NoteName?> = arrayOf(NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b)
    private val bassStandardNotes: Array<NoteName?> = arrayOf(NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c, NoteName.d, NoteName.e, NoteName.f, NoteName.g, NoteName.a, NoteName.b, NoteName.c, NoteName.d)

    inner class NotePosn(internal var noteName: NoteName?, internal var yLower: Float, internal var yUpper: Float, internal var octave: Int)

    private val ALTO_POSN = ArrayList<NotePosn>()

    private fun getNotePosn(y: Float, clef: Clef): ArrayList<NotePosn> {
        var iterArray : Array<NoteName?> = arrayOfNulls<NoteName>(9)
        var res: ArrayList<NotePosn> = ArrayList()
        when (clef) {
            Clef.ALTO -> {
                iterArray = altoStandardNotes
                res = ALTO_POSN
            }
            Clef.TREBLE -> {
                iterArray = trebleStandardNotes
                res = ALTO_POSN
                Octave.octave = 5
            }
            Clef.BASS -> {
                iterArray = bassStandardNotes
                res = ALTO_POSN
                Octave.octave = 4
            }
        }
        for (i in 14 downTo -1 + 1) {
            val yL = interval * 8 - subInt - subInt * (i + 1)
            val yU = interval * 8 - subInt * (i + 1)
            var o = octave
            if (clef == Clef.ALTO) {
                if (i < 7) {
                    if (i == 0) {
                        o = octave - 2
                    } else {
                        o = octave - 1
                    }
                } else if (i == 14) {
                    o = octave + 1
                }
            } else if (clef == Clef.TREBLE) {
                if (i < 8) {

                    o = octave - 1
                } else if (i == 0) {
                    o = octave - 2
                }
            } else if (clef == Clef.BASS) {
                if (i < 6) {
                    o = octave - 2
                } else if (i < 13 && i >= 6) {
                    o = octave - 1
                }

            }

            res.add(NotePosn(iterArray[i], yL, yU, o))
        }
        return res
    }

    private fun getNoteFromPosn(y: Float, clef: Clef): NoteName? {
        val noteList = getNotePosn(y, clef)
        for (i in noteList.indices) {
            if (y > noteList[i].yLower && y <= noteList[i].yUpper) {
                octave = noteList[i].octave
                return noteList[i].noteName
            }
        }
        octave += 1
        return NoteName.c
    }


    private fun accidental(note: NoteName, accidental: Int): NoteName {
        if (accidental == 1) {
            for (sharpNote in SHARPS) {
                if (note == sharpNote[0] && note != NoteName.b && note != NoteName.e) {
                    return sharpNote[1]
                }
            }
        }
        if (accidental == -1) {
            for (flatNote in FLATS) {
                if (note == flatNote[0]) {
                    return flatNote[1]
                }
            }
        }
        return note
    }

    private fun keyAccidental(note: NoteName): NoteName {
        if (Key.COUNT == 0) {
            return note
        }
        if (Key.SHARP) {

            for (i in 0..Key.COUNT - 1) {
                if (note == SHARPS[i][0]) {
                    accidental = 1
                    return SHARPS[i][1]

                }
            }
        } else {
            for (i in 0..Key.COUNT - 1){
                if (note == FLATS[i][0]){
                    accidental = -1
                    return FLATS[i][1]

                }
            }
        }
        return note
    }


    init {
        y = snapNoteY(yC)
        name = accidental(keyAccidental(getNoteFromPosn(snapNoteY(yC) - DensityMetrics.getToolbarHeight(), ClefSetting.clef)!!), acc)
        rhythm = LastRhythm.value

    }


    //Change this to a rounding statement
    private fun snapNoteY(y: Float): Float {

        var snapY = 0f
        for (i in 0..12) {
            if (interval * i - subSubInt < y && interval * i + subSubInt >= y) {
                snapY = interval * i + DensityMetrics.getToolbarHeight()
                break
            } else if (interval * i + subInt - subSubInt < y && interval * i + subInt + subSubInt >= y) {
                snapY = interval * i + subInt + DensityMetrics.getToolbarHeight()
                break
            }
        }
        return snapY
    }


}
